File: src/ru/edu/vtisov/array/binarysearch/FindFirstAndLastPositionOfElementInSortedArray.java
package ru.edu.vtisov.array.binarysearch;

public class FindFirstAndLastPositionOfElementInSortedArray {
    public int[] searchRange(int[] nums, int target) {
        int first = binarySearch(nums, target, true);
        int second = binarySearch(nums, target, false);
        return new int[] {first, second};
    }

    private int binarySearch(int[] nums, int target, boolean isLeft) {
        int result = -1;
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] == target) {
                result = mid;
                if (isLeft) end = mid - 1;
                else start = mid + 1;
            } else if (nums[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return result;
    }
}

File: src/ru/edu/vtisov/array/binarysearch/find-first-and-last-position-of-element-in-sorted-array.go
package binarysearch

func searchRange(nums []int, target int) []int {
	left := binarySearch(nums, target, true)
	right := binarySearch(nums, target, false)
	return []int{left, right}
}

func binarySearch(nums []int, target int, isLeft bool) int {
	i, j := 0, len(nums)-1
	for i <= j {
		mid := i + ((j - i) / 2)
		if nums[mid] == target {
			if isLeft {
				if (mid == 0) || nums[mid-1] != target {
					return mid
				}
				j = mid - 1
			} else {
				if (mid == len(nums)-1) || nums[mid+1] != target {
					return mid
				}
				i = mid + 1
			}
		} else if nums[mid] < target {
			i = mid + 1
		} else {
			j = mid - 1
		}
	}
	return -1
}
